zap/src/config.rs

```diff
@@ -324,10 +347,34 @@ impl<'src> Ty<'src> {
 			// note the lack of - 1 here, it's because we need to store 0 length as well.
 			Ty::Enum(Enum::Unit(variants)) => Some(NumTy::from_f64(0.0, variants.len() as f64)),
 			Ty::Enum(Enum::Tagged { variants, .. }) => Some(NumTy::from_f64(0.0, variants.len() as f64)),
-			Ty::Ref(.., variants_size) => *variants_size,
+			Ty::Ref(.., ty) => ty.variants_size(),
 			_ => None,
 		}
 	}
+
+	pub fn primitive_ty(&self) -> PrimitiveTy<'src> {
+		match self {
+			Ty::Num(..) => PrimitiveTy::Name("number"),
+			Ty::Str(..) => PrimitiveTy::Name("string"),
+			Ty::Buf(..) => PrimitiveTy::Name("buffer"),
+			Ty::BrickColor => PrimitiveTy::Name("BrickColor"),
+			Ty::DateTimeMillis => PrimitiveTy::Name("DateTime"),
+			Ty::DateTime => PrimitiveTy::Name("DateTime"),
+			Ty::Boolean => PrimitiveTy::Name("boolean"),
+			Ty::Color3 => PrimitiveTy::Name("Color3"),
+			Ty::Vector2 => PrimitiveTy::Name("Vector2"),
+			Ty::Vector3 => PrimitiveTy::Name("Vector3"),
+			Ty::Vector(..) => PrimitiveTy::Name("Vector3"),
+			Ty::AlignedCFrame => PrimitiveTy::Name("CFrame"),
+			Ty::CFrame => PrimitiveTy::Name("CFrame"),
+			Ty::Instance(class) => PrimitiveTy::Instance(*class),
+			Ty::Enum(Enum::Unit(variants)) => PrimitiveTy::Unit(variants.clone()),
```

Comment:
Tagged enums should be primative? As we can differentiate between them at runtime?

Reply:
They're checked by their variants, and if a variant overlaps a duplicate type error is reported.